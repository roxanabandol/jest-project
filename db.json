{
  "forms": [
    {
      "id": 11,
      "volunteerName": "Roxana",
      "hours": 1,
      "jobTitle": "raisingFunds"
    },
    {
      "id": 9,
      "volunteerName": "Test",
      "hours": 2,
      "jobTitle": "foodSupply"
    },
    {
      "id": 3,
      "volunteerName": "Jack",
      "hours": 3,
      "jobTitle": "adoptionPromotion"
    }
  ]
}