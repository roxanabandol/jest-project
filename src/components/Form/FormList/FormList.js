import React from "react";

const FormList = ({ forms }) => (
  <table className="center">
    <thead>
      <tr>
        <th>Name</th>
        <th>Hours available</th>
        <th>Job</th>
      </tr>
    </thead>
    <tbody>
      {(forms || []).map((form, formIndex) => (
        <tr key={formIndex}>
          <td>{form.volunteerName}</td>
          <td>{form.hours}</td>
          <td>{form.jobTitle}</td>
        </tr>
      ))}
    </tbody>
  </table>
);

export default FormList;
