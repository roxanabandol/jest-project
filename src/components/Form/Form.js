import React, { useState, useEffect } from "react";
import axios from "axios";
import styled from "styled-components";

import FormList from "./FormList/FormList";

const StyledForm = styled.div`
  form {
    width: 300px;
    margin: 0 auto;

    .group {
      display: grid;
      text-align: left;
    }
  }

  table,
  th,
  td {
    border: 1px solid black;
  }
  table.center {
    margin-left: auto;
    margin-right: auto;
  }
`;

const jobs = ["Raising funds", "Food supply", "Adoption promotion"];

function initialState() {
  return { volunteerName: "", hours: "", jobTitle: jobs[0] };
}

function Form() {
  const [form, setForm] = useState(initialState());
  const [forms, setForms] = useState([]);

  const { volunteerName, hours, jobTitle } = form;

  useEffect(() => {
    axios
      .get("http://localhost:3000/forms")
      .then((response) => {
        setForms(response.data);
      })
      .catch((error) => {
        console.log(error);
      });
  }, [forms]);

  function updateField(key, event) {
    setForm({ ...form, [key]: event.target.value });
  }

  function resetForm() {
    setForm(initialState);
  }

  function handleSubmit(event) {
    event.preventDefault();

    const { volunteerName, hours, jobTitle } = form;

    const newPost = {
      id: Math.floor(Math.random() * 100 + 1),
      volunteerName,
      hours: parseInt(hours),
      jobTitle,
    };

    axios
      .post("http://localhost:3000/forms", newPost)
      .then(() => {
        setForm({ ...newPost });
      })
      .catch((error) => {
        console.log(error);
      });
  }

  return (
    <StyledForm>
      <div className="post-form">
        <form>
          <div className="group">
            <label>Name:</label>
            <input
              id="volunteerName-input"
              name="volunteerName"
              type="text"
              value={volunteerName}
              onChange={(event) => updateField("volunteerName", event)}
            />
          </div>
          <div className="group">
            <label>Number of hours available per week:</label>
            <input
              id="hours-input"
              name="hours"
              type="number"
              value={hours}
              onChange={(event) => updateField("hours", event)}
            />
          </div>

          <div className="group">
            <label>Availability on one area:</label>
            <select
              id="job-input"
              name="jobTitle"
              value={jobTitle}
              onChange={(event) => updateField("jobTitle", event)}
            >
              {jobs.map((job, jobIndex) => (
                <option key={jobIndex} value={job}>
                  {job}
                </option>
              ))}
            </select>
          </div>

          <button
            type="submit"
            id="submit-button"
            onClick={(event) => handleSubmit(event)}
          >
            Submit
          </button>
          <button type="button" id="reset-button" onClick={resetForm}>
            Reset
          </button>
        </form>
        <FormList forms={forms} />
      </div>
    </StyledForm>
  );
}

export default Form;
