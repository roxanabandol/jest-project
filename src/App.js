import React from "react";
import styled from "styled-components";

import Form from "./components/Form/Form";

const StyledApp = styled.div`
  background-color: white;
  height: 100%;
  padding: 20px;
  text-align: center;
`;

function App() {
  return (
    <StyledApp>
      <h1>Become a volunteer</h1>
      <Form />
    </StyledApp>
  );
}

export default App;
