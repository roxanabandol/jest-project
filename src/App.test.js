import React from "react";
import { shallow, mount, configure } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

import App from "./App";
import Form from "./components/Form/Form";
import FormList from "./components/Form/FormList/FormList";

import { forms } from "../db.json";

configure({ adapter: new Adapter() });

const simulateOnChangeInput = (wrapper, inputSelector, newValue) => {
  const input = wrapper.find(inputSelector);
  input.simulate("change", {
    target: { value: newValue },
  });

  return wrapper.find(inputSelector);
};

describe("The components are rendered", () => {
  it("renders App component without crashing", () => {
    shallow(<App />);
  });

  it("renders Form component without crashing", () => {
    shallow(<Form />);
  });

  it("renders FormList component without crashing", () => {
    shallow(<FormList />);
  });

  it("renders title without crashing", () => {
    const wrapper = shallow(<App />);

    const header = <h1>Become a volunteer</h1>;
    expect(wrapper.contains(header)).toBe(true);
  });

  it("renders form inputs", () => {
    const wrapper = shallow(<Form />);

    expect(wrapper.find('input[name="volunteerName"]')).toHaveLength(1);
    expect(wrapper.find('input[name="hours"]')).toHaveLength(1);
    expect(wrapper.find('select[name="jobTitle"]')).toHaveLength(1);
  });

  it("renders submit button without crashing", () => {
    const wrapper = shallow(<Form />);

    const label = wrapper.find("#submit-button").text();
    expect(label).toBe("Submit");
  });
});

describe("FormList is passing props", () => {
  it("accepts props", () => {
    const wrapper = mount(<FormList forms={forms} />);
    expect(wrapper.props().forms).toEqual(forms);
  });

  it(`should have all properties`, () => {
    const wrapper = mount(<FormList forms={forms} />);
    const renderedForms = wrapper.props().forms;
    for (let i = 0; i < renderedForms.length; i += 1) {
      expect(renderedForms[i]).toHaveProperty("id");
      expect(renderedForms[i]).toHaveProperty("volunteerName");
      expect(renderedForms[i]).toHaveProperty("hours");
      expect(renderedForms[i]).toHaveProperty("jobTitle");
    }
  });
});

describe("The events are working ", () => {
  it("fill the form with values and then reset the form", () => {
    const wrapper = shallow(<Form />);

    const updatedNameInput = simulateOnChangeInput(
      wrapper,
      "#volunteerName-input",
      "Endzela Gaye"
    );
    const updatedHourInput = simulateOnChangeInput(
      wrapper,
      "#hours-input",
      "2"
    );
    const updatedJobTitleInput = simulateOnChangeInput(
      wrapper,
      "select",
      "Food supply"
    );

    expect(updatedJobTitleInput.props().value).toBe("Food supply");
    expect(updatedHourInput.props().value).toBe("2");
    expect(updatedNameInput.props().value).toBe("Endzela Gaye");

    wrapper.find("#reset-button").simulate("click");

    expect(wrapper.find("#volunteerName-input").props().value).toBe("");
    expect(wrapper.find("#hours-input").props().value).toBe("");
    expect(wrapper.find("#job-input").props().value).toBe("Raising funds");
  });

  it("fill the form with values and then submit the form", () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow(<Form />);

    simulateOnChangeInput(wrapper, "#volunteerName-input", "Endzela Gaye");
    simulateOnChangeInput(wrapper, "#hours-input", "2");
    simulateOnChangeInput(wrapper, "select", "Food supply");

    const button = shallow(
      <button
        type="submit"
        id="submit-button"
        className="submit"
        onClick={mockCallBack}
      >
        Submit
      </button>
    );
    button.find("#submit-button").simulate("click", {
      preventDefault: () => {},
    });
    expect(mockCallBack).toHaveBeenCalledTimes(1);
  });
});

describe("Snapshot", () => {
  it("matches App the snapshot", () => {
    const wrapper = mount(<App />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
