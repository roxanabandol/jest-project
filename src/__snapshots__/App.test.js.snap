// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches App the snapshot 1`] = `
<App>
  <styled.div>
    <div
      className="sc-gsTCUz fYSMQF"
    >
      <h1>
        Become a volunteer
      </h1>
      <Form>
        <styled.div>
          <div
            className="sc-bdfBwQ hiYNTf"
          >
            <div
              className="post-form"
            >
              <form>
                <div
                  className="group"
                >
                  <label>
                    Name:
                  </label>
                  <input
                    id="volunteerName-input"
                    name="volunteerName"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="group"
                >
                  <label>
                    Number of hours available per week:
                  </label>
                  <input
                    id="hours-input"
                    name="hours"
                    onChange={[Function]}
                    type="number"
                    value=""
                  />
                </div>
                <div
                  className="group"
                >
                  <label>
                    Availability on one area:
                  </label>
                  <select
                    id="job-input"
                    name="jobTitle"
                    onChange={[Function]}
                    value="Raising funds"
                  >
                    <option
                      key="0"
                      value="Raising funds"
                    >
                      Raising funds
                    </option>
                    <option
                      key="1"
                      value="Food supply"
                    >
                      Food supply
                    </option>
                    <option
                      key="2"
                      value="Adoption promotion"
                    >
                      Adoption promotion
                    </option>
                  </select>
                </div>
                <button
                  id="submit-button"
                  onClick={[Function]}
                  type="submit"
                >
                  Submit
                </button>
                <button
                  id="reset-button"
                  onClick={[Function]}
                  type="button"
                >
                  Reset
                </button>
              </form>
              <FormList
                forms={Array []}
              >
                <table
                  className="center"
                >
                  <thead>
                    <tr>
                      <th>
                        Name
                      </th>
                      <th>
                        Hours available
                      </th>
                      <th>
                        Job
                      </th>
                    </tr>
                  </thead>
                  <tbody />
                </table>
              </FormList>
            </div>
          </div>
        </styled.div>
      </Form>
    </div>
  </styled.div>
</App>
`;
